cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -0,0 +1,90 @@
+
+        // Best pattern to use here?
+        // - Store in structs in Self along the way?
+        // - Type States (like the old publish?)
+        // - Make each phase independant, taking arguments for all that's needed
+        //    and track the ongoing state only here in publish()?
+        // - Combination of Type States and making each phase independant may be the 
+        //    easiest to test?
+        prepublish_check();
+
+        prepublish_build();
+
+        gather_build_info();
+
+        publish_to_catalog();
+    }
+
+    /// Ensures current commit is in remote
+    /// Gathers of the info needed for publish
+    /// ... locked_url, rev, rev_count, rev_date of the build repo from git
+    /// ... locked_url, rev, rev_count, rev_date of the base catalog from the
+    ///         lockfile of the build repo
+    /// ... version, description, name, attr_path from manifest
+    /// ... catalog from ???? (cli argument? manifest?)
+    pub fn prepublish_check(repo_dir: &Path, package_name: &str)
+    {
+        // _sandbox = temp dir
+        // Use GitCommandProvider to get remote and current rev of build_repo
+        // Use GitCommandProvider to clone that remote/rev to _sandbox
+            // this will ensure it's in the remote
+        // Load the manifest from the .flox of that repo to get the version/description/catalog
+    }
+    
+    /// Access a `flox build` builder, probably another provider?
+    /// ... to clone into a sandbox and run
+    /// ... `flox activate; flox build;`
+    pub fn prepublish_build<B: ManifestBuilder>(builder: &B, _sandbox: &Path)
+    {
+        // We need to do enough to get the build info for the next step
+        // ... equivalent to run `flox activate; flox build;`?
+
+        // This will create a flox env in the _sandbox, right? so should we pass that back
+        // and store it to use for the next step?
+
+        // Use ManifestBuilder to perform the build (evaluation?) so we can get
+        // the info we need.  Do we need to build? or is there something else?
+        let build_output = builder.build(_sandbox, flox_env, &self.package_name);
+        
+        // the build_output looks to be the stdout of the build process.. how do 
+        // we access the drv_path and that stuff?
+    }
+
+    /// Obtains info from the build process (the sandboxed one above)
+    /// ... base_catalog_{locked_url, rev, rev_count, rev_date}
+    /// ... system, drv_path (??), outputs (??)
+    pub fn gather_build_info(_sandbox: &Path)
+    {
+        // Access lockfile of _sandbox to get base_catalog_{locked_url, rev, rev_count, rev_date}
+        // populate the following:
+        //      how do we get the drv_path, outputs(and paths), system from this?
+    }
+
+    /// Uses client catalog to...
+    /// ... check access to the catalog
+    /// ... check presence of and create the package if needed
+    /// ... publish the build info
+    pub fn publish_to_catalog()
```

Comment:
We'll need to come up with a recovery strategy if uploading the metadata succeeds and uploading the artifact fails. Some kind of `invalidate_package_metadata` function? But then you need to ensure the user/client has the correct permissions to _only_ revoke the package they just published and not every package in their catalog. Tricky.

Reply:
This highlights the issue that "publish" is doing two publishes at once, i.e. publish metadata and publish binaries, metadata is valid without the binaries (but not vice versa). So technically failure at uploading the binaries is still a partial success, although you'd then need a way to upload binaries separateley  (or generally do both steps separately if you wanted to)

I think we had something like this in mind some time before, to ensure publish always publishes both:

- write publish metadata (marked tentative)
- copy closure
- confirm metadata

if unconfirmed the metadata is never published and deleted after a while.
This _should_ be less tricky than revoking and would mean that metadata is correctly unpublished in the case:

- publish metadata
- ⚡️ network outage ⚡️
- publish binary  fails
- **revoking fails**

Reply:
I don't think it's clear, at least not to me, yet that this would necessarily be a failure.  if the first cut line of useful functionality doesn't need the binary store, that implies the metadata itself in the catalog is sufficient to serve a purpose on it's own.  If the upload fails, then the  user will have to build it, but it's still useful and I don't think we need to garbage collect.  Just an opposing idea that still needs some discussion.

Reply:
We discussed last week, but my take on this is that:

* regardless if anything else, we need garbage collection of the binary cache
    * the "aliveness" of any artifact is determined by the existence of a reference to a package (and its subsequent closure) in the catalog
* if we upload the packages first, and then the subsequent metadata submission fails, then there are three possible outcomes:
    1. if the package is re-published before GC runs then that subsequent publish will be faster because the package closure is already there
    2. if GC runs before the package is re-published then the subsequent publish is exactly as fast as it would have been had it been successful the first time
    3. if that particular version of the package is never re-published then there is no record of it ever having been published in the catalog, which is _probably_ what the user would prefer

So on this basis I'd say the simple approach to this problem is 1) upload to binary cache first, then 2) upload the metadata, and 3) let GC handle everything else.